core/src/prepare.rs

```diff
@@ -0,0 +1,252 @@
+    /// [SourceCache]. Note that this function will not load standard library modules into
+    /// the cache or process them if they are not already loaded.
+    pub fn from_stdlib<C: Cache>(cache: &mut SourceCache, eval_cache: &mut C) -> Self {
+        let stdlib_terms = stdlib_terms(cache);
+
+        let type_ctxt = typecheck::mk_initial_ctxt(stdlib_terms.as_slice()).unwrap();
+
+        let mut eval_env = Environment::new();
+        for (module, key) in cache.stdlib_modules().iter().copied() {
+            // The internals module needs special treatment: it's required to be a record
+            // literal, and its bindings are added directly to the environment
+            if let stdlib::StdlibModule::Internals = module {
+                let result = eval::env_add_record(
+                    eval_cache,
+                    &mut eval_env,
+                    Closure::atomic_closure(
+                        cache
+                            .term_owned(key)
+                            .expect("The standard library is parsed whenever it is loaded"),
+                    ),
+                );
+                if let Err(eval::EnvBuildError::NotARecord(rt)) = result {
+                    panic!(
+                        "InitialEnvs::from_stdlib: \
+                        expected the stdlib module {} to be a record, got {:?}",
+                        cache.source_path(key),
+                        rt
+                    )
+                }
+            } else {
+                eval::env_add(
+                    eval_cache,
+                    &mut eval_env,
+                    module.name().into(),
+                    cache
+                        .term_owned(key)
+                        .expect("The standard library is parsed whenever it is loaded"),
+                    Environment::new(),
+                )
+            }
+        }
+
+        Self {
+            eval_env,
+            type_ctxt,
+        }
+    }
+}
+
+impl Default for InitialEnvs {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// Parse a source and populate the corresponding entry in the cache, or do
+/// nothing if the entry has already been parsed.
+pub fn parse(cache: &mut SourceCache, cache_key: CacheKey) -> Result<ParseErrors, ParseErrors> {
+    match cache.get(cache_key) {
+        SourceState::Added => {
+            let (term, parse_errors) = parse_multi(cache, cache_key, InputFormat::default())?;
+            cache.record_parse_errors(cache_key, parse_errors.clone());
+            cache.set_parsed(
+                cache_key,
+                ParsedEntry {
+                    term,
+                    state: ParsedState::Parsed,
+                },
+            );
+            Ok(parse_errors)
+        }
+        _ => Ok(cache
+            .get_parse_errors(cache_key)
+            .expect("Any parsed entry should have a corresponding entry in parse_errors")
```

Comment:
Should we encore that with an enum, instead of a separate `EntryState` in a struct?

Reply:
I'm not sure why I wanted the parse errors to be stored separately from the cache entries, but that seems silly to me now. I've changed the code to store parse errors directly in the cache entry for parsed entries. The only wrinkle is that it would be annoying to keep parse errors around after the entry is advanced from "parsed" to "typechecked". But would there be any reason to proceed to typechecking for a source file that doesn't parse correctly, even in the LSP?

Reply:
Ah, I believe we do proceed with typechecking on a source file that doesn't parse correctly. If only because typechecking is actually the driver for the whole code analysis (goto(ref|def), hover, etc.), and we want to provide analysis for as much code as possible. In fact the typechecker explicitly ignore parse errors in the AST for that very reason:  https://github.com/tweag/nickel/blob/baf6e3d3c1d2361c75e927c908081b9042cf5f77/core/src/typecheck/mod.rs#L1797